c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -541,6 +549,10 @@ pub fn visit<'tcx>(
         tc.constraints.add_all_perms(ptr, perms);
     }
 
+    for (ptr, perms) in acx.gacx.known_fn_ptr_perms() {
+        tc.constraints.add_all_perms(ptr, perms);
+    }
```

Comment:
This is O(N^2), since `dataflow::type_check::visit` runs on (roughly) every function, and `known_fn_ptr_perms` visits every function to check for `KnownFn` permissions.

First, I don't think it's necessary to use constraints to set the permissions on these pointers.  The `AllPerms` constraint just ORs some bits into the `PointerId`'s entry in the `Assignment`, which is easy to do directly without using `DataflowConstraints`.  I think we could set these permissions up front, then run the dataflow analysis, and finally assert that the permissions were unchanged during the analysis (mainly, if the program is demanding more of a libc function's return value than it actually provides, that's a problem).

Second, if we do actually need constraints here, it would be more efficient to build a global baseline `DataflowConstraints` and clone it for each `visit`, rather than reconstructing those constraints each time.

Reply:
> First, I don't think it's necessary to use constraints to set the permissions on these pointers. The `AllPerms` constraint just ORs some bits into the `PointerId`'s entry in the `Assignment`, which is easy to do directly without using `DataflowConstraints`.

Ah, I didn't realize we could do that.  That would definitely be better.  What's the API for that?

> finally assert that the permissions were unchanged during the analysis (mainly, if the program is demanding more of a libc function's return value than it actually provides, that's a problem).

Like we do for the string literal permissions, right?  That makes sense.

> Second, if we do actually need constraints here, it would be more efficient to build a global baseline `DataflowConstraints` and clone it for each `visit`, rather than reconstructing those constraints each time.

Yeah, I could definitely do that, though I'd prefer a better way to set the permissions in the first place.

Reply:
> First, I don't think it's necessary to use constraints to set the permissions on these pointers. The `AllPerms` constraint just ORs some bits into the `PointerId`'s entry in the `Assignment`, which is easy to do directly without using `DataflowConstraints`.

I'm still unsure how to do this, so for now I did this:

> Second, if we do actually need constraints here, it would be more efficient to build a global baseline `DataflowConstraints` and clone it for each `visit`, rather than reconstructing those constraints each time.

in 84d8be42363d9ca5ce25878287e3400f261f4372.

Reply:
Nevermind, I think I might've figured out how to do that.  Let me try that way.

Reply:
> First, I don't think it's necessary to use constraints to set the permissions on these pointers. The `AllPerms` constraint just ORs some bits into the `PointerId`'s entry in the `Assignment`, which is easy to do directly without using `DataflowConstraints`. I think we could set these permissions up front, then run the dataflow analysis, 

Done in 5c093c687ca001ad5f13c319aec0f0888fe85b3b.

> and finally assert that the permissions were unchanged during the analysis (mainly, if the program is demanding more of a libc function's return value than it actually provides, that's a problem).

Done in c5444e9c17ee248869b7cc41fff464bae50346b3.

Reply:
Yep, that's what I was thinking of